File: examples/bench/mutual_recursion_compare.f
# Mutual recursion benchmark (call overhead + branchless recursion).

ping(0, acc): acc.
ping(n, acc): pong(n - 1, acc + 1).

pong(0, acc): acc.
pong(n, acc): ping(n - 1, acc + 1).

run_ping(0, _, acc): acc.
run_ping(k, n, acc): run_ping(k - 1, n, acc + ping(n, 0)).

main():
    n <- 800
    rounds <- 2000
    say(run_ping(rounds, n, 0))
    1.

File: examples/bench/mutual_recursion_compare.py
#!/usr/bin/env python3

import sys


def ping(n: int, acc: int) -> int:
    if n == 0:
        return acc
    return pong(n - 1, acc + 1)


def pong(n: int, acc: int) -> int:
    if n == 0:
        return acc
    return ping(n - 1, acc + 1)


if __name__ == "__main__":
    sys.setrecursionlimit(100000)
    n = 800
    rounds = 2000
    total = 0
    for _ in range(rounds):
        total += ping(n, 0)
    print(total)
